pkg/signaling/signaler.go

```diff
@@ -0,0 +1,63 @@
+package signaling
+
+import (
+	"context"
+	"sync"
+)
+
+type ctxKey[T Signal] struct {
+	Signal T
+}
+
+type Signal interface {
+	error
+	comparable
+}
+
+type signaler = chan struct{}
```

Comment:
nit; a new type is slightly better than an alias here; the make(signaler) will work the same, but it discourages users from making some assumptions

```suggestion
type signaler chan struct{}
```